src/node_webstorage.cc

```diff
@@ -0,0 +1,654 @@
+#include "sqlite3.h"
+
+namespace node {
+namespace webstorage {
+
+using v8::Array;
+using v8::Boolean;
+using v8::Context;
+using v8::DontDelete;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::IndexedPropertyHandlerConfiguration;
+using v8::Integer;
+using v8::Isolate;
+using v8::Local;
+using v8::Map;
+using v8::Maybe;
+using v8::MaybeLocal;
+using v8::Name;
+using v8::NamedPropertyHandlerConfiguration;
+using v8::Null;
+using v8::Object;
+using v8::PropertyCallbackInfo;
+using v8::PropertyDescriptor;
+using v8::PropertyHandlerFlags;
+using v8::String;
+using v8::Uint32;
+using v8::Value;
+
+static constexpr const int kMaxStorageSize = 10 * 1024 * 1024;  // 10 MB
+
+#define THROW_SQLITE_ERROR(env, r)                                            \
+  node::THROW_ERR_INVALID_STATE((env), sqlite3_errstr((r)))
+
+#define CHECK_ERROR_OR_THROW(env, expr, expected, ret)                        \
+  do {                                                                        \
+    int _r = (expr);                                                          \
+    if (_r != (expected)) {                                                   \
+      THROW_SQLITE_ERROR((env), _r);                                          \
+      return (ret);                                                           \
+    }                                                                         \
+  } while (0)
+
+static void ThrowQuotaExceededException(Local<Context> context) {
+  Isolate* isolate = context->GetIsolate();
+  auto dom_exception_str = FIXED_ONE_BYTE_STRING(isolate, "DOMException");
+  auto err_name = FIXED_ONE_BYTE_STRING(isolate, "QuotaExceededError");
+  auto err_message = FIXED_ONE_BYTE_STRING(isolate,
+      "Setting the value exceeded the quota");
+  Local<Object> per_context_bindings;
+  Local<Value> domexception_ctor_val;
+  if (!GetPerContextExports(context).ToLocal(&per_context_bindings) ||
+      !per_context_bindings->Get(context, dom_exception_str)
+          .ToLocal(&domexception_ctor_val)) {
+    return;
+  }
+  CHECK(domexception_ctor_val->IsFunction());
+  Local<Function> domexception_ctor = domexception_ctor_val.As<Function>();
+  Local<Value> argv[] = {err_message, err_name};
+  Local<Value> exception;
+
+  if (!domexception_ctor->NewInstance(context, arraysize(argv), argv)
+           .ToLocal(&exception)) {
+    return;
+  }
+
+  isolate->ThrowException(exception);
+}
+
+Storage::Storage(
+    Environment* env, Local<Object> object, Local<String> location) :
+    BaseObject(env, object) {
+  MakeWeak();
+  node::Utf8Value utf8_location(env->isolate(), location);
+  _symbols.Reset(env->isolate(), Map::New(env->isolate()));
+  _db = nullptr;
+  _location = utf8_location.ToString();
+}
+
+Storage::~Storage() {
+  sqlite3_close(_db);
+  _db = nullptr;
+}
+
+void Storage::MemoryInfo(MemoryTracker* tracker) const {
+  tracker->TrackField("symbols", _symbols);
+}
+
+bool Storage::Open() {
```

Comment:
@cjihrig: I suppose my key question is whether the sqllite dependency is actually *needed* to support this API.  I get that it makes certain parts of the implementation easier but the API does not surface any of the actual capabilities of sqllite to the user. Would a non-sqllite implementation under the covers be more trouble that it is worth? Are there particular reasons to stay with sqllite?